### PR TITLE
Add ISensor.Update()

### PR DESCRIPTION
**Pull request by chriselion**
This add an ISensor.Update() method and implements it for the cases where it makes sense. This prevents observations from being cleared before they can be sent to the DemonstrationRecorder.

### CODE REVIEW

File: UnitySDK/Assets/ML-Agents/Editor/Tests/DemonstrationTests.cs [hunk 1]
```diff
@@ -57,5 +60,73 @@ public void TestStoreInitalize()
...
+            var agentEnableMethod = typeof(Agent).GetMethod("OnEnable",
+                BindingFlags.Instance | BindingFlags.NonPublic);
+            var agentSendInfo = typeof(Agent).GetMethod("SendInfo",
+                BindingFlags.Instance | BindingFlags.NonPublic);
+
+            agentEnableMethod?.Invoke(agent1, new object[] { });
+            academyInitializeMethod?.Invoke(aca, new object[] { });
+
+            // Step the agent
+            agent1.RequestDecision();
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Editor/Tests/DemonstrationTests.cs]: Unfortunately this doesn't fail without the fix in place. Might need to discuss a better way to get the test Agent to act like it's in inference mode.

File: UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs [hunk 2]
```diff
@@ -35,15 +36,16 @@ void Update()
         /// <summary>
         /// Creates demonstration store for use in recording.
         /// </summary>
-        void InitializeDemoStore()
+        public void InitializeDemoStore(IFileSystem fileSystem = null)
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs]: Allow for the mock file system to be injected. Could keep this private and do it by reflection, but not sure if that's necessary.

File: UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs [hunk 3]
```diff
@@ -71,14 +73,22 @@ public void WriteExperience(AgentInfo info)
             m_DemoStore.Record(info);
         }
 
+        public void Close()
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs]: Should this happen on disable instead? Need it for testing to make sure the write stream is flushed,

File: UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs [hunk 4]
```diff
@@ -21,12 +22,14 @@ public class DemonstrationStore
 
         public DemonstrationStore(IFileSystem fileSystem)
         {
-            m_FileSystem = fileSystem;
-        }
-
-        public DemonstrationStore()
-        {
-            m_FileSystem = new FileSystem();
+            if (fileSystem != null)
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/DemonstrationStore.cs]: Combined these.

File: UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs (hunk 3 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs]: would you move the initialization to OnEnable?

Comment (chriselion) [reply UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs]: It's already done lazily in Update()

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/DemonstrationRecorder.cs]: should it be set to `null` after it's closed?  Is there a use case for stopping recording during play mode and resuming?

**Pull request merged**